### ISSUE TITLE
Rhetorical: Why did caskroom just re-install (rather than migrate) every app in Brewfile?

### ISSUE THREAD

**Issue by remy-actual** [caskroom/homebrew-cask#21894]
Yes, I know __why__. But since a main topic of discussion in "the relevant [issue](https://github.com/caskroom/homebrew-cask/issues/21603) and [pull request](https://github.com/caskroom/homebrew-cask/pull/21857)" is the Caskroom UX, I thought I'd share my User Experience this morning.  

Steps to replicate: 
* Run Brewfile
* Encounter questions

1. Why is cask reporting Adium not currently installed? 
1. We're re-installing Adium now? Okay...
1. Now cask is reporting BitBar is not currently installed too? Um... I'm _looking at it_ running in the Menu Bar right now.
1. Did I mess something up? Is Cask misconfigured? 
1. Are we re-installing everything? A - Z ?
1. 😑  We **are** reinstalling everything. Greeeeeaaat. How long is this going to take?
1. "It seems there is already an App at '/Applications/iTerm.app'; not moving." No kidding! Where do you think I'm reading this? 
1. So... does this mean I have two copies of iTerm on my system now? Two Google Chromes? 
1. How much am I going to have to clean up? Am I going to have a bunch of broken links? Working links pointing to multiple installs? This is turning into a real PITA. 
1. How much of this is running `brew cask cleanup` going to fix?
1. 😱  Are my app preferences being preserved? Or are they getting wiped out? Because _that would **suck**_. Guess we'll wait and see! 
1. Well my custom key maps aren't functioning now. Time to go re-setup Karabiner! 😡 
1. Ugh. MacTeX. Seriously? All I wanted was to install colordiff and brew upgrade. I'm gonna go take a shower... 
1. Okay. 30 minutes and 6.6 gigs later and we're done! Glad I have a decent processor and unlimited reliable bandwidth. Feel bad today for anybody who doesn't 😕 ...
1. Ugh. What is this system pop-up? I don't even know what app is sending this... 
![whoareyou](https://cloud.githubusercontent.com/assets/952821/15992860/af1f77b6-308b-11e6-81e0-e468016655b4.png)
1. Is this the thing they were talking about in the README about changing Caskroom behavior? They wouldn't push out a major change like that without including **some** notification on the command line, right? I mean, Cask _is_ a CLI workflow -- _why wouldn't you advise your users of changes in the CLI?_
1. Maybe they did and I just didn't see it because I'm calling Brewfile from a script. But surely that's a very common use case for Cask, right? 

Not offering this as a complaint. Just a head's up for how pushing out silent changes can really bork someone's work day.  

Anyways... gonna go play detective in my filesystem now.

**Comment by vitorgalvao** [comment]
@remy-actual Thank you for your well-written, detailed, understanding, and polite comment. It’s a good example to any bug reporter. Since I like it a lot, I’ll make this issue the default to explain this problem, and what we’re doing. First, the thought process<sup id="a1">[1](#f1)</sup>:

+ We are alpha-grade software, and you **should not yet rely on us to be stable and not break things**. This has been made clear and repeated since the inception of the software.
+ `brew cask list` is broken and you **should not rely on it under any under circumstance**. This has been made clear and repeated for over one year.
+ There is no upgrade functionality, and **any upgrading tricks you use are unsupported**. This has been made clear and repeated for over one year.
+ We will undergo under major breaking changes, **and are aware things will break**. [This has been made clear and repeated for almost a year](https://github.com/caskroom/homebrew-cask/issues/13201).
+ The alternative to breaking things is doing nothing and stagnation. **This is not supposition, but a proven reality**. We know because the project was dead core-wise for a long time.

All that said, the project was stagnant before the addition of @jawshooah and the return of @claui, both of which we’re very lucky to have (and [every other maintainer](https://github.com/orgs/caskroom/people), naturally, but they are particularly relevant to this point since they work a lot on the core and the core was the stagnant part).

In addition, I did not predict this would cause such distress to some people, seeing as [the much bigger behavioural change](https://github.com/caskroom/homebrew-cask/pull/13966) has been going pretty smoothly, all things considered. I also [wanted this new change to be made really soon](https://github.com/caskroom/homebrew-cask/pull/21857#issue-159776160) so it could piggyback on the other pull request and cause the least amount of distress (everything at once, instead of one major break here, one major break there).

So in summation, my apologies for the disturbance. [A fix is on the way](https://github.com/caskroom/homebrew-cask/issues/21858#issuecomment-225452879).

I ask that any non-maintainer commenting after this only does so with something constructively relevant. Insults and comments about what we should have done<sup id="a2">[2](#f2)</sup> will lead us nowhere, and will be deleted<sup id="a3">[3](#f3)</sup>. This comment admits the mistake, gives context as to why it happened, and describes the coming fix. Please do not comment if your comment isn’t pertinent to the fix.

Thank you in advance for understanding and keeping things civil.

---

<sup id="f1">[1](#a1)</sup> If it seems like I’m evading blame, rest assured I am not, as you will see if you read until the end.
<sup id="f2">[2](#a2)</sup> Especially from users who never otherwise contributed, even in discussions, as is the norm.
<sup id="f3">[3](#a3)</sup> It pains me a bit to do so, but I will not submit myself or the other maintainers to abuse (though it has yet to happen on this issue, I don’t see it far off).

**Comment by carlosefr** [comment]
It's understandable that, as any other open-source project, this is all done in a best-effort basis. So, when thinks break, we get to keep both pieces. No problem with that. Not a point in complaining about that.

However, the project does have a website, which is the first and last thing that many/most users will ever  read about the project. Yet there isn't a single mention about it being "alpha-grade" there. Nor are there any mention of important features being known to be broken. User expectations are built on what they can easily find, not on what's dispersed throughout GitHub PR discussions or other such places, which they will most likely never see.

Breaking user expectations and then arguing that things "have been repeated and made clear for over one year" is not a proper argument when most users created those expectations for material that clearly misrepresents the state of the project.

At the very least, http://caskroom.io should have easy to find information about known issues and breaking changes.

**Comment by vitorgalvao** [comment]
> However, the project does have a website, which is the first and last thing that many/most users will ever read about the project.

You are absolutely speculating. You have no data or reason to claim that as fact. Just as easily I’ll say that most users find the github page first and only, and I have a huge reason to believe that: frequently the webpage had wrong information for a good while until anyone noticed.

Also, I do believe you haven’t read or paid attention to my post. I made exceedingly clear I wasn’t evading blame when making the top points, and still you picked on those.

I also asked people to not comment if not related to the fix. This is done mostly to prevent the issue being drawn out by discussion that isn’t relevant to arriving at a resolution.

I also made it clear that comments about what we should have done from users that never participate where unwelcome in this matter. Quite frankly, I’m tired of answering those users, as they have no context about the decisions, and the same points must be made over and over and over.

We’re all volunteers, here, an owe nothing to anyone. If a user won’t bother searching a bit and finding a bit more about the project and previous decisions before commenting, we shouldn’t be forced to spend our time repeating the same things.

As such, and as you did not follow my simple and clear request, I’m sad to say I’ll make the first comment removal.

Your comment added nothing of value, and all your points were addressed in my previous one. Please do not reply again unless you are addressing the fix.

**Comment by carlosefr** [comment]
Although on the internet everyone seems to be angry, I was absolutely not assigning any blame nor was I  criticizing the effort you (and others) invest in the project.

Managing user expectations is an important part of managing a community. Not allowing discussion around these issues and considering it unconstructive criticism worthy of removal is your choice.

**Comment by vitorgalvao** [comment]
> Managing user expectations is an important part of managing a community.

Yes, we know. This project has existed for a few years, we understand what it means to have a community. [With all due respect](https://youtu.be/lQECNWI4RHI), you are not part of the community. You are a user, you never participated before in discussions of the project.

> Not allowing discussion around these issues and considering it unconstructive criticism worthy of removal is your choice.

I’m not stifling discussion on this issue. I’ve made my intentions pretty clear: I do not want to drawn out discussion with irrelevant things. And yes, your comments are very unconstructive, since **they added nothing to what had already been said**. It is not a discussion when you’re just reiterating points I’ve already admitted, replied to, and offered the resolution.

We know we screwed up (mostly I did). I explained very well in which ways the screw up was made, and what is being done about it.

Every comment you make and make any one on the team waste time answering, it’s time being wasted that could be spent fixing stuff.

Now, please, go to sleep. It’s already late where we live and I don’t intend to keep discussing this. Unless it’s related to the fix, it does not matter.

**Comment by vitorgalvao** [comment]
Since in less than an hour I’ve already been forced to play whack-a-mole and am starting to get unfriendly twitter mentions, I’ve locked the issue.

My sincere apologies to everyone that has a genuine comment related to the fix and now is prevented from commenting, but I’m a person, I had a full day, and I am tired and just want to go to sleep. Frankly, had it not been for seeing an email related to this and seeing some people were having issues, I wouldn’t even have opened my laptop after I got home. I did so because I care about this community and our users, and you deserve a reassurance we’re taking care of this. More specifically, at the moment @jawshooah is. It was [his comment](https://github.com/caskroom/homebrew-cask/issues/21858#issuecomment-225452879) that alerted me.

Every maintainer of this project, past and present, has been a pleasure to work with, and I’m really glad to have them all on this team. Most of our contributors, experienced and unexperienced alike, are nice and understanding people that I like to encounter every day, even if to many I simply merge the PR without saying a word. Rest assured, those are a huge favourite to merge, when everything is fine. We simply get a huge barrage of daily PRs, and thanking you all individually every time does add up.

When I read a comment, I always read it in the most positive way possible. I always assume the user is typing it with a smile and being friendly. That way only the unambiguously aggressive comments are treated as such. Those are a shame when they happen, and even more of a shame when they continue after all the points have been addressed in a polite manner.

Occasionally we do encounter some entitled users, or users that just want to argue. Those users quite often get me close to burnout. Those users are also always users that haven’t previously commented or contributed. Those users do not care that we are all people and volunteers, doing the best we can with the time available. Those users do not take the time to understand a bit of our history and the “why” of decisions, the context of an always evolving project. Those users, to put it bluntly, are not worth it. Those users can bugger off; their absence is appreciated.

And then we have users like @remy-actual, which even in frustration can make a friendly bug report, not bashing the people that caused the frustration, but commenting on the cause itself. Those are the ones worth it. Those many times end up being great maintainers as well (I’ll mention @adidalal as one of the friendliest maintainers I’ve ever encountered, in any project).

When maintainers burnout, the whole community suffers. I do not want to burnout, I want to make this project the best it can be, even if it means breaking things and owning that breakage, or breaking things and scaling back the breakage.

Thank you to all of you that make it worth it. Thank you to all of you who are friendly and understand healthy people make healthy software.
